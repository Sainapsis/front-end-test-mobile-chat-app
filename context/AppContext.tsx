import React, { createContext, useContext, ReactNode } from 'react';
import { useUser } from '@/hooks/useUser';
import { useChats } from '@/hooks/useChats';
import { DatabaseProvider } from '../database/DatabaseProvider';
import { useDatabase } from '@/hooks/useDatabase';
import { User } from '@/types/User';
import { Chat } from '@/types/Chat';

/**
 * Interface representing the application context state
 */
type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  /** List of chats for the current user */
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (chatId: string, text: string, senderId: string, imageUri?: string) => Promise<boolean>;
  loading: boolean;
  dbInitialized: boolean;
  clearChats: (userId: string) => Promise<void>;
  deleteChat?: (chatId: string, userId: string) => Promise<void>;
  deleteMessage?: (messageId: string, chatId: string) => Promise<boolean>;
  addReaction?: (messageId: string, emoji: string) => Promise<boolean>;
  removeReaction?: (reactionId: string, messageId: string) => Promise<boolean>;
  editMessage?: (messageId: string, newText: string) => Promise<boolean>;
};

/**
 * Context for managing application-wide state
 */
const AppContext = createContext<AppContextType | undefined>(undefined);

/**
 * Internal component that wraps the application content with context
 * @param children - Child components to be wrapped by the context
 */
function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabase();
  const userContext = useUser();
  const chatContext = useChats(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value: AppContextType = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
    deleteChat: chatContext.deleteChat || undefined,
    deleteMessage: chatContext.deleteMessage || undefined,
    addReaction: chatContext.addReaction || undefined,
    removeReaction: chatContext.removeReaction || undefined,
    editMessage: chatContext.editMessage || undefined,
    sendMessage: async (chatId: string, text: string, senderId: string, imageUri?: string) => {
      if (!chatContext.sendMessage) return false;
      return chatContext.sendMessage(chatId, text, senderId, imageUri);
    },
  };

  if (!isInitialized) {
    return null;
  }

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

/**
 * Provider component that wraps the application with necessary contexts
 * @param children - Child components to be wrapped by the provider
 */
export const AppProvider = ({ children }: { children: React.ReactNode }) => (
  <DatabaseProvider>
    <AppContent>{children}</AppContent>
  </DatabaseProvider>
);

/**
 * Hook for accessing application context
 * @returns Application context containing user, chat, and database states
 * @throws Error if used outside of AppProvider
 */
export function useAppContext() {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error('useAppContext must be used within an AppProvider');
  }
  return context;
}
