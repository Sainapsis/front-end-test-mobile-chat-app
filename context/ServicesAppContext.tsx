import React, { createContext, useContext, ReactNode } from 'react';
import { DatabaseProvider, useDatabaseStatus } from './DataBaseContext';
import { useChatsDb,  useUserDb } from '../hooks/db';
import { Chat, Message, User } from '@/utils/types';

type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (
    chatId: string, 
    text: string | undefined, 
    senderId: string,
    mediaUrl?: string,
    mediaType?: string,
    mediaThumbnail?: string,
    voiceUrl?: string,
    voiceDuration?: number,
    isVoiceMessage?: boolean,
    
  ) => Promise<Message | null>;
  addReaction: (messageId: string, userId: string, emoji: string) => Promise<boolean>;
  removeReaction: (reactionId: string) => Promise<boolean>;
  searchMessages: (query: string, chatId?: string) => Promise<void>;
  searchResults: Message[];
  deleteMessage: (messageId: string) => Promise<boolean>;
  editMessage: (messageId: string, newText: string) => Promise<boolean>;
  updateMessageStatus: (messageId: string, status: 'sent' | 'delivered' | 'read') => Promise<void>;
  markMessagesAsRead: (chatId: string) => Promise<void>;
  loading: boolean;
  dbInitialized: boolean;
};

export const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabaseStatus();
  const userContext = useUserDb();
  const chatContext = useChatsDb(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export function AppProvider({ children }: { children: ReactNode }) {
  return (
    <DatabaseProvider>
      <AppContent>{children}</AppContent>
    </DatabaseProvider>
  );
}

