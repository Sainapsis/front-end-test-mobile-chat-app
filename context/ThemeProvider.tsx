import React, { useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import ThemeContext from './ThemeContext';
import { Theme } from '@/types/tColores';

/**
 * Provider component that handles theme state and persistence
 * @param children - Child components to be wrapped by the provider
 */
export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>('light');

  useEffect(() => {
    loadTheme();
  }, []);

  const loadTheme = async () => {
    try {
      const savedTheme = await AsyncStorage.getItem('theme');
      if (savedTheme) {
        setTheme(savedTheme as Theme);
      }
    } catch (error) {
      console.error('Error loading theme:', error);
    }
  };

  const toggleTheme = async () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    try {
      await AsyncStorage.setItem('theme', newTheme);
    } catch (error) {
      console.error('Error saving theme:', error);
    }
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};