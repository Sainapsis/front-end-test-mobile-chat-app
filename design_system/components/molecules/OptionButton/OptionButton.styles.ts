import { StyleSheet } from 'react-native';
import { spacing, typography } from '@/design_system/ui/tokens';

export const styles = StyleSheet.create({
  optionButton: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: spacing.sm,
    width: '100%',
    paddingHorizontal: spacing.sm,
  },
  optionText: {
    fontSize: typography.sizes.md,
    fontFamily: typography.families.primary,
    fontWeight: typography.weights.regular,
    lineHeight: typography.lineHeights.normal,
    marginLeft: spacing.sm,
  },
});