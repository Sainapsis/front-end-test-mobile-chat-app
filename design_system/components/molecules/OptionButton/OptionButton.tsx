import React from 'react';
import { Pressable, Text } from 'react-native';
import Ionicons from 'react-native-vector-icons/Ionicons';
import { styles } from './OptionButton.styles';
import { themes } from '@/design_system/ui/tokens';
import { useColorScheme } from '@/hooks/useColorScheme';

interface OptionButtonProps {
  /** Name of the icon to be displayed */
  icon: string;
  /** Text to be displayed next to the icon */
  text: string;
  /** Function to be called when the button is pressed */
  onPress: () => void;
  /** Custom color for the text */
  color?: string;
}

/**
 * OptionButton component provides a button with an icon and text.
 * It is used to present options or actions to the user in a visually consistent way.
 */
export const OptionButton: React.FC<OptionButtonProps> = ({ icon, text, onPress, color = 'black'}) => {
  const theme = useColorScheme() ?? 'light';
  return (
    <Pressable style={styles.optionButton} onPress={onPress} android_ripple={{ color: '#ddd', borderless: false }}>
      <Ionicons name={icon} size={20} color={ theme === 'light' ? themes.light.text.contrast : themes.dark.text.contrast } />
      <Text style={[styles.optionText, { color }]}>{text}</Text>
    </Pressable>
  );
}

export default OptionButton;