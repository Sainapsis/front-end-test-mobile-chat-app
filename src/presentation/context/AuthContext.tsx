import { LoginParams } from "@/src/data/interfaces/auth.interface";
import { createContext, ReactNode, useContext } from "react";
import { useAuth } from "../hooks/useAuth";
import { User } from "@/src/domain/entities/user";

interface AuthContextType {
  currentUser: User | null;
  login: ({ userId }: LoginParams) => Promise<boolean>;
  logout: () => void;
}

const _AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthContext({ children }: { children: ReactNode }) {
  const authContext = useAuth();

  const value = {
    ...authContext,
  };

  return (
    <_AuthContext.Provider value={value}>{children}</_AuthContext.Provider>
  );
}

export function useAuthContext() {
  const context = useContext(_AuthContext);

  if (context === undefined) {
    throw new Error("useAuthContext must be used within an AppProvider");
  }

  return context;
}
