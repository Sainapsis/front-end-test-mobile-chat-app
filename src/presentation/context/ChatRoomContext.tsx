import { DeleteMessageParams, EditMessageParams, UpdateStatusMessageParams } from '@/src/data/interfaces/chatRoom.interface';
import { Chat } from '@/src/domain/entities/chat';
import { Message, MessageStatus } from '@/src/domain/entities/message';
import { createContext, ReactNode, useContext } from "react";
import { useChatRoom } from '../hooks/useChatRoom';

interface ChatRoomContextType {
  updateStatus: (
    currentUserId: string,
    chat: Chat,
    statusMessage: MessageStatus
  ) => Promise<void>;
  sendMessage: (chatId: string, message: Message) => Promise<boolean>;
  editMessage: ({
    chatId,
    messageId,
    newText,
  }: EditMessageParams) => Promise<boolean>;
  deleteMessage: ({
    chatId,
    messageId,
  }: DeleteMessageParams) => Promise<boolean>;
  updateMessageStatus: (
    chatId: string,
    { messageId, status }: UpdateStatusMessageParams
  ) => Promise<void>;
  handleLoadMoreMessage: ({ chatId }: { chatId: string }) => Promise<void>;
}

const _ChatRoomContext = createContext<ChatRoomContextType | undefined>(undefined);

export function ChatRoomContext({ chatId, children }: { chatId: string; children: ReactNode }) {
  const chatRoomContext = useChatRoom({ chatId });

  const value = {
    ...chatRoomContext,
  };

  return (
    <_ChatRoomContext.Provider value={value}>{children}</_ChatRoomContext.Provider>
  );
}

export function useChatRoomContext() {
  const context = useContext(_ChatRoomContext);

  if (context === undefined) {
    throw new Error("useChatRoomContext must be used within an AppProvider");
  }

  return context;
}
