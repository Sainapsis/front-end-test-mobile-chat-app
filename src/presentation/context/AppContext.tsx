import React, { createContext, useContext, ReactNode } from "react";
import { useDatabase } from "../hooks/useDatabase";
import { useUser } from "@/src/presentation/hooks/useUser";
import { DatabaseContext } from "@/src/presentation/context/DatabaseProvider";
import { UserContext } from "./UserContext";
import { AuthContext } from "./AuthContext";

interface AppContextType {
  dbInitialized: boolean;
}

const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabase();
  const userContext = useUser();
  
  const value = {
    ...userContext,
    dbInitialized: isInitialized,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export function AppProvider({ chatId, children }: { chatId: string; children: ReactNode }) {
  return (
    <DatabaseContext>
      <AuthContext>
        <UserContext>
          <AppContent>{children}</AppContent>
        </UserContext>
      </AuthContext>
    </DatabaseContext>
  );
}

export function useAppContext() {
  const context = useContext(AppContext);

  if (context === undefined) {
    throw new Error("useAppContext must be used within an AppProvider");
  }

  return context;
}
