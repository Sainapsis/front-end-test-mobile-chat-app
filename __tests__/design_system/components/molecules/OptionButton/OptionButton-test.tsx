import * as React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '@/context/ThemeContext';
import { OptionButton } from '@/design_system/components/molecules/OptionButton/OptionButton';

// Mock para Ionicons
jest.mock('react-native-vector-icons/Ionicons', () => {
    const React = require('react');
    const { View } = require('react-native');
  
    return {
      __esModule: true,
      default: ({ name, size, color }: { name: string; size: number; color: string }) => (
        <View testID={`icon-${name}`} style={{ width: size, height: size, backgroundColor: color }} />
      ),
    };
  });
  
  

// Mock para useColorScheme
jest.mock('@/hooks/useColorScheme', () => ({
  useColorScheme: jest.fn(() => 'light')
}));

describe('OptionButton Component', () => {
  const defaultProps = {
    icon: 'heart',
    text: 'Like',
    onPress: jest.fn()
  };

  const renderWithTheme = (component: string | number | boolean | Iterable<React.ReactNode> | React.JSX.Element | null | undefined) => {
    return render(
      <ThemeProvider>
        {component}
      </ThemeProvider>
    );
  };

  it('renders correctly with snapshot', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <OptionButton {...defaultProps} />
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with correct text', () => {
    const { getByText } = renderWithTheme(
      <OptionButton {...defaultProps} />
    );
    expect(getByText('Like')).toBeTruthy();
  });

  it('calls onPress when pressed', () => {
    const onPressMock = jest.fn();
    const { getByText } = renderWithTheme(
      <OptionButton {...defaultProps} onPress={onPressMock} />
    );
    
    fireEvent.press(getByText('Like'));
    expect(onPressMock).toHaveBeenCalledTimes(1);
  });

  it('renders with custom color', () => {
    const customColor = '#FF0000';
    const { getByText } = renderWithTheme(
      <OptionButton {...defaultProps} color={customColor} />
    );
    
    expect(getByText('Like')).toBeTruthy();
  });

  it('renders with different icons', () => {
    const icons = ['heart', 'share', 'chatbubble', 'bookmark'];
    
    icons.forEach(icon => {
      const { getByText } = renderWithTheme(
        <OptionButton {...defaultProps} icon={icon} text={`${icon} action`} />
      );
      expect(getByText(`${icon} action`)).toBeTruthy();
    });
  });
});