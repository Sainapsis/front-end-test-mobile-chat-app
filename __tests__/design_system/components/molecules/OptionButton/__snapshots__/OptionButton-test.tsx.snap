// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionButton Component renders correctly with snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "paddingHorizontal": 8,
      "paddingVertical": 8,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#FFFFFF",
        "height": 20,
        "width": 20,
      }
    }
    testID="icon-heart"
  />
  <Text
    style={
      [
        {
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 24,
          "marginLeft": 8,
        },
        {
          "color": "black",
        },
      ]
    }
  >
    Like
  </Text>
</View>
`;
